probe-rs/src/probe/espusbjtag/protocol.rs

```diff
@@ -0,0 +1,613 @@
+use core::fmt;
+use std::{fmt::Debug, time::Duration};
+
+use rusb::{request_type, Context, Device, Direction, TransferType, UsbContext};
+
+use crate::{
+    DebugProbeError, DebugProbeInfo, DebugProbeSelector, DebugProbeType, ProbeCreationError,
+};
+
+const JTAG_PROTOCOL_CAPABILITIES_VERSION: u8 = 1;
+const JTAG_PROTOCOL_CAPABILITIES_SPEED_APB_TYPE: u8 = 1;
+const MAX_COMMAND_REPETITIONS: usize = 1024;
+const OUT_BUFFER_SIZE: usize = OUT_EP_BUFFER_SIZE * 64;
+const OUT_EP_BUFFER_SIZE: usize = 64;
+const IN_EP_BUFFER_SIZE: usize = 64;
+const USB_TIMEOUT: Duration = Duration::from_millis(5000);
+const USB_DEVICE_CLASS: u8 = 0xFF;
+const USB_DEVICE_SUBCLASS: u8 = 0xFF;
+const USB_DEVICE_PROTOCOL: u8 = 0x01;
+const USB_DEVICE_TRANSFER_TYPE: TransferType = TransferType::Bulk;
+
+const USB_CONFIGURATION: u8 = 0x0;
+
+const USB_VID: u16 = 0x303A;
+const USB_PID: u16 = 0x1001;
+
+const VENDOR_DESCRIPTOR_JTAG_CAPABILITIES: u16 = 0x2000;
+
+// const VENDOR_JTAG_SETDIV: u16 = 0;
+// const VENDOR_JTAG_SETIO: u16 = 1;
+// const VENDOR_JTAG_GETTDO: u16 = 2;
+
+// const VENDOR_JTAG_SETIO_TDI: u16 = 1 << 0;
+// const VENDOR_JTAG_SETIO_TMS: u16 = 1 << 1;
+// const VENDOR_JTAG_SETIO_TCK: u16 = 1 << 2;
+// const VENDOR_JTAG_SETIO_TRST: u16 = 1 << 3;
+// const VENDOR_JTAG_SETIO_SRST: u16 = 1 << 4;
+
+pub(super) struct ProtocolHandler {
+    // The USB device handle.
+    device_handle: rusb::DeviceHandle<rusb::Context>,
+
+    // The command in the queue and their repetitions.
+    // For now we do one command at a time.
+    command_queue: Option<(Command, usize)>,
+    output_buffer: Vec<Command>,
+    input_buffer: Vec<u8>,
```

Comment:
Please add comments to explain what `out_buffer`, `input_buffer` and `buffer` are used for. `input_buffer` seems to be unused?

Reply:
I had a mixup there. I removed `buffer` and moved all usages to `input_buffer`. I also added docs.